File: 5 Dec Wall Colouring II/code.py
class Solution:
    def minCost(self, costs : list[list[int]]) -> int:
        # code here
        n = len(costs)
        m = len(costs[0])
        if(m==1):
            return costs[0][0] if n==1 else -1
        for i in range(1, n):
            r = sorted(enumerate(costs[i-1]), key=lambda p: p[1])
            for k in range(m):
                if r[0][0] != k:
                    costs[i][k] += r[0][1]
                else:
                    costs[i][k] += r[1][1]
        return min(costs[-1])

File: 16 Dec Strings Rotations of Each Other/code.cpp
class Solution {
  public:
    bool areRotations(string &s1, string &s2) {
        // code here
        string t="";
        for(auto c:s1)t+=c;
        for(auto c:s1)t+=c;
        int n=s2.size();
        vector<int>lp(n,0);
        int i=1,j=0;
        while(i<n){
            if(s2[i]==s2[j])lp[i++]=++j;
            else{
                if(j)j=lp[j-1];
                else i++;
            }
        }
        i=0,j=0;
        while(i<(n+n)){
            if(t[i]==s2[j]){i++,j++;if(j==n)return true;}
            else{
                if(j)j=lp[j-1];
                else i++;
            }
        }
        return false;
    }
};